### PR TITLE
Fix trappy timeouts in data stream APIs

### PR DESCRIPTION
**Pull request by DaveCTurner**
Relying on the default 30s timeout is trappy, we should be explicit
about the timeouts we're using in these requests.

Relates #107984

### CODE REVIEW

**Comment by DaveCTurner** [comment]
This was all very mechanical, mostly just using IntelliJ's extract-parameter refactoring.

File: modules/data-streams/src/main/java/org/elasticsearch/datastreams/rest/RestModifyDataStreamsAction.java [hunk 1]
```diff
@@ -40,13 +39,18 @@ public List<Route> routes() {
     protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
         ModifyDataStreamsAction.Request modifyDsRequest;
         try (XContentParser parser = request.contentParser()) {
-            modifyDsRequest = ModifyDataStreamsAction.Request.PARSER.parse(parser, null);
+            modifyDsRequest = ModifyDataStreamsAction.Request.PARSER.parse(
+                parser,
+                actions -> new ModifyDataStreamsAction.Request(
+                    RestUtils.getMasterNodeTimeout(request),
+                    RestUtils.getAckTimeout(request),
+                    actions
```
Comment (DaveCTurner) [inline modules/data-streams/src/main/java/org/elasticsearch/datastreams/rest/RestModifyDataStreamsAction.java:47]: No need to mutate the request, can just create it with all the right properties up-front.

File: modules/data-streams/src/main/java/org/elasticsearch/datastreams/rest/RestPromoteDataStreamAction.java [hunk 2]
```diff
@@ -31,7 +32,10 @@ public List<Route> routes() {
 
     @Override
     protected RestChannelConsumer prepareRequest(RestRequest restRequest, NodeClient client) throws IOException {
-        PromoteDataStreamAction.Request request = new PromoteDataStreamAction.Request(restRequest.param("name"));
+        PromoteDataStreamAction.Request request = new PromoteDataStreamAction.Request(
+            RestUtils.getMasterNodeTimeout(restRequest),
```
Comment (DaveCTurner) [inline modules/data-streams/src/main/java/org/elasticsearch/datastreams/rest/RestPromoteDataStreamAction.java:36]: API change: no longer hard-coded

**Review by DaveCTurner — COMMENTED** [review]
Specific changes:

1. Pull out `masterNodeTimeout` and `ackTimeout` parameters on all request constructor methods.
2. Pass in `TEST_REQUEST_TIMEOUT` for both timeouts in all tests.
3. Use `RestUtils` to extract timeouts in REST APIs.
4. Add REST params where they were missing, and update `rest-api-spec`.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DeleteStep.java [hunk 3]
```diff
@@ -52,7 +53,10 @@ public void performDuringNoSnapshot(IndexMetadata indexMetadata, ClusterState cu
                 // This is the last backing index in the data stream, and it's being deleted because the policy doesn't have a rollover
                 // phase. The entire stream needs to be deleted, because we can't have an empty list of data stream backing indices.
                 // We do this even if there are multiple failure store indices because otherwise we would never delete the index.
-                DeleteDataStreamAction.Request deleteReq = new DeleteDataStreamAction.Request(new String[] { dataStream.getName() });
+                DeleteDataStreamAction.Request deleteReq = new DeleteDataStreamAction.Request(
+                    MasterNodeRequest.infiniteMasterNodeTimeout(currentState.getMinTransportVersion()),
```
Comment (DaveCTurner) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DeleteStep.java:57]: Slight behaviour change, previously we'd time out after 30s here, but waiting forever seems more appropriate.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/AnalyticsCollectionService.java [hunk 4]
```diff
@@ -81,7 +82,11 @@ public void putAnalyticsCollection(
         assert (state.nodes().isLocalNodeElectedMaster());
 
         AnalyticsCollection collection = new AnalyticsCollection(request.getName());
-        CreateDataStreamAction.Request createDataStreamRequest = new CreateDataStreamAction.Request(collection.getEventDataStream());
+        CreateDataStreamAction.Request createDataStreamRequest = new CreateDataStreamAction.Request(
+            TimeValue.THIRTY_SECONDS /* TODO should we wait longer? */,
+            TimeValue.THIRTY_SECONDS /* TODO should we wait longer? */,
```
Comment (DaveCTurner) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/AnalyticsCollectionService.java:87]: Unsure about these timeouts tbh, I'm leaving them at 30s but at least calling them out as such here.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/AnalyticsCollectionService.java [hunk 5]
```diff
@@ -124,7 +129,10 @@ public void deleteAnalyticsCollection(
         assert (state.nodes().isLocalNodeElectedMaster());
 
         AnalyticsCollection collection = new AnalyticsCollection(request.getCollectionName());
-        DeleteDataStreamAction.Request deleteDataStreamRequest = new DeleteDataStreamAction.Request(collection.getEventDataStream());
+        DeleteDataStreamAction.Request deleteDataStreamRequest = new DeleteDataStreamAction.Request(
+            TimeValue.THIRTY_SECONDS /* TODO should we wait longer? */,
```
Comment (DaveCTurner) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/AnalyticsCollectionService.java:133]: Unsure about this timeout tbh, I'm leaving it at 30s but at least calling it out as such here.

File: x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/Fleet.java [hunk 6]
```diff
@@ -300,6 +301,7 @@ public void cleanUpFeature(ClusterService clusterService, Client client, ActionL
         if (dataStreamDescriptors.isEmpty() == false) {
             try {
                 Request request = new Request(
+                    TimeValue.THIRTY_SECONDS /* TODO should we wait longer? */,
```
Comment (DaveCTurner) [inline x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/Fleet.java:304]: Unsure about this timeout tbh, I'm leaving it at 30s but at least calling it out as such here.

File: x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/persistence/ProfilingDataStreamManager.java [hunk 7]
```diff
@@ -141,8 +141,11 @@ public void onFailure(Exception e) {
     private void createDataStream(ProfilingDataStream dataStream, final ActionListener<? super ActionResponse> listener) {
         final Executor executor = threadPool.generic();
         executor.execute(() -> {
-            CreateDataStreamAction.Request request = new CreateDataStreamAction.Request(dataStream.getName());
-            request.masterNodeTimeout(TimeValue.timeValueMinutes(1));
+            CreateDataStreamAction.Request request = new CreateDataStreamAction.Request(
+                TimeValue.ONE_MINUTE /* TODO should we wait longer? */,
+                TimeValue.THIRTY_SECONDS /* TODO should we wait longer? */,
```
Comment (DaveCTurner) [inline x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/persistence/ProfilingDataStreamManager.java:146]: Unsure about these timeouts tbh, I'm leaving them as-is but at least calling them out as such here.

**Comment by DaveCTurner** [comment]
Serverless failure is unrelated to this change - it does compile, and that's the only danger here.

**Review by arteam — APPROVED** [review]
The PR looks good to me from the API perspective,  but I'm also not sure whether 30s timeouts are sufficient for the Fleet tests, it would great to get another opinion from the Fleet team.

**Comment by DaveCTurner** [comment]
> I'm also not sure whether 30s timeouts are sufficient for the Fleet tests

This PR doesn't change that, it just makes it explicit.

**Pull request merged**